ibmcloud: check worker node version

Pacho20 opened the pull request:
**- Description of the problem which is fixed/What is the use case**
This fixes issues where worker nodes run a different OCP version than the control plane. This can cause issues with the DaemonSet installation mode.
**- What I did**
Add a function to check the cloud provider in the DaemonSet install script. If the provider is ibmcloud, retrieve the node's version and use it to get the extension image for that specific version instead of using the environment variable.

Pacho20 committed:
ibmcloud: check worker node version

Add a function to check the cloud provider in the DaemonSet install
script. If the provider is ibmcloud, retrieve the node's version and
use it to get the extension image for that specific version instead of
using the environment variable. This fixes issues where worker nodes
run a different OCP version than the control plane.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -187,9 +187,66 @@ uninstall() {
+		echo "Can't find the CLOUD_PROVIDER in the peer-pods-cm"
+		exit 1
+	fi
+	
```
Please drop the leading tab.

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -187,9 +187,66 @@ uninstall() {
 }
 
+get_cloud_provider() {
+	provider="$(kubectl get configmap/peer-pods-cm -n openshift-sandboxed-containers-operator -o jsonpath='{.data.CLOUD_PROVIDER}')"
```
It is a bit unfortunate to make `peer-pods-cm` a requirement for this script. What about getting the information from a `CLOUD_PROVIDER` environment variable ? This can be easily provided by the controller.

Pacho20 replied:
True, but it would require a new release of the operator. If we plan to move the entire logic to the operator anyway, can we address this at that point and make this PR a quick fix? If you think I should pass the provider as an environment variable now I can do that.

gkurz replied:
That's fine for the temporary quick fix.

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -187,9 +187,66 @@ uninstall() {
+	raw_version="$(kubectl get nodes/${NODE_NAME} -o jsonpath='{.metadata.labels.ibm-cloud\.kubernetes\.io/worker-version}')"
+
+	if [[ -z "${raw_version}" ]]; then
+		echo "Can't find the ibm-cloud.kubernetes.io/worker-version label"
```
Errors should go to stderr. Since this script has quite some instances of that, I'd suggest to introduce an `error()` helper that writes to stderr and does `exit 1` (in a preparatory commit).

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -187,9 +187,66 @@ uninstall() {
+
 	mkdir -p "/usr/share/rpm-ostree/extensions"
-	extract_container_image "$EXTENSION_IMAGE" "/usr/share/rpm-ostree/extensions" "/usr/share/rpm-ostree" "/tmp/regauth/auth.json"
+	extract_container_image "$extension_image" "/usr/share/rpm-ostree/extensions" "/usr/share/rpm-ostree" "/tmp/regauth/auth.json"
```
This basically means that the OSC controller generates an incorrect `EXTENSION_IMAGE` in the IBM Cloud case and we have to override it here. It is acceptable for a quick fix but for the long term, I'd prefer the controller to implement the appropriate logic directly.

gcoon151 replied:
agree this is short term fix to get it working again and we'll longer term move this to better places.

Pacho20 replied:
I agree; I marked the function with a FIXME. This PR is intended as a quick fix, and we already have plans to implement this in the operator.

gkurz replied:
Works for me.

Pacho20 committed:
daemonset deployment: add error helper to lib.sh

Introduce a helper function in lib.sh for kata-install scripts to
standardize error handling:

- error: log an error message to stderr and return 1

This helper will replace ad-hoc error handling in later commits.

Signed-off-by: Patrik Fodor <patrik.fodor@ibm.com>

bpradipt commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -187,9 +187,63 @@ uninstall() {
+	echo "${provider}"
+}
+
+get_worker_node_version_ibmcloud() {
```
This appears to be sane logic to detect the version of the worker node for retrieving the correct extension image.
Why this has to be provider specific? Is this due to the non-availability of the annotation in other providers ?

Pacho20 replied:
Yes, that's the main reason. I don't know if other providers have a similar label or annotation, but I can make the function non-provider-specific if you think that's a better approach. I didn’t really consider this because it’s intended as a quick fix, and the logic will be moved into the operator. See [Greg's comment](https://github.com/openshift/sandboxed-containers-operator/pull/1468#discussion_r2607644857).

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -187,9 +187,112 @@ uninstall() {
 }
 
+# get_cloud_provider OUT_VAR
+# Writes the cloud provider string to OUT_VAR.
```
This shell construct is far too unusual and it obfuscates the code for no benefice. Please revert to do the usual way, i.e. return result with `echo`.

Pacho20 replied:
done

gkurz commented on scripts/kata-install/lib.sh:
```diff
@@ -55,3 +55,17 @@ client_tools() {
+}
+
+# Fatal: log and exit
+die() {
```
All the user scripts are started with `set -e` => `return 1` in `error()` is already fatal.

`die()` isn't needed in this context and it doesn't have any user in this series. Please drop it for the sake of simplicity. The `Non-fatal/Fatal` comments should be removed as well.

Pacho20 replied:
Done

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -187,9 +187,102 @@ uninstall() {
+		-n openshift-sandboxed-containers-operator \
+		-o jsonpath='{.data.CLOUD_PROVIDER}')"; then
+		error "get_cloud_provider: kubectl failed to query peer-pods-cm"
+		return
```
`return` is dead code since `set -e` will cause `error()` to fail. Please remove it and other occurences.

Pacho20 replied:
As I understand it, return is necessary when using command substitution. Functions executed via command substitution run in a subshell, and the subshell does not inherit set -e.
Without return, the script continues execution even after an error occurs.

Some logs running this without `return`:
```
...
++ get_cloud_provider
++ local provider
+++ kubectl get configmap/peer-pods-cm -n openshift-sandboxed-containers-operator -o 'jsonpath={.data.CLOUD_PROVIDER}'
++ provider=
++ [[ -z '' ]]
++ error 'get_cloud_provider: CLOUD_PROVIDER not set in configmap/peer-pods-cm'
++ local 'msg=get_cloud_provider: CLOUD_PROVIDER not set in configmap/peer-pods-cm'
++ printf 'ERROR: %s\n' 'get_cloud_provider: CLOUD_PROVIDER not set in configmap/peer-pods-cm'
ERROR: get_cloud_provider: CLOUD_PROVIDER not set in configmap/peer-pods-cm
++ return 1
++ echo ''
+ provider=
+ case "${provider}" in
+ extension_image=quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:e40ac6498a409397d37c563f377675408a871f9971d354bb8f1a0b1845a0381f
+ mkdir -p /usr/share/rpm-ostree/extensions
+ extract_container_image quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:e40ac6498a409397d37c563f377675408a871f9971d354bb8f1a0b1845a0381f /usr/share/rpm-ostree/extensions /usr/share/rpm-ostree /tmp/regauth/auth.json
...

the script continues "normally"
```

Some logs running with `return`:
```
...
++ get_cloud_provider
++ local provider
+++ kubectl get configmap/peer-pods-cm -n openshift-sandboxed-containers-operator -o 'jsonpath={.data.CLOUD_PROVIDER}'
++ provider=
++ [[ -z '' ]]
++ error 'get_cloud_provider: CLOUD_PROVIDER not set in configmap/peer-pods-cm'
++ local 'msg=get_cloud_provider: CLOUD_PROVIDER not set in configmap/peer-pods-cm'
++ printf 'ERROR: %s\n' 'get_cloud_provider: CLOUD_PROVIDER not set in configmap/peer-pods-cm'
ERROR: get_cloud_provider: CLOUD_PROVIDER not set in configmap/peer-pods-cm
++ return 1
++ return
+ provider=

script exits here, Pod goes into error state
```

The other solutions I found for this problem seemed more complicated than simply using return.

gkurz replied:
Oh yes you're right. Let's keep that then.

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -187,9 +187,102 @@ uninstall() {
+	fi
+
+	echo "${provider}"
+	return 0
```
Not needed. Please drop it and other occurences.

Pacho20 replied:
Done

gkurz reviewed (commented):
Thanks for the changes @Pacho20 . Some more polishing and it should be good to go.

gkurz commented:
/ok-to-test

gkurz reviewed (approved):
/lgtm

Thanks @Pacho20 !
